cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/service/AuthorizationDatasetTypeService.java

```diff
@@ -0,0 +1,177 @@
+  }
+
+  @Override
+  public List<DatasetModuleMeta> listModules(final NamespaceId namespaceId) throws Exception {
+    List<DatasetModuleMeta> modules = delegate.listModules(namespaceId);
+
+    Principal principal = authenticationContext.getPrincipal();
+    List<DatasetModuleMeta> visibleModules = new ArrayList<>(modules.size());
+    for (List<DatasetModuleMeta> split : Lists.partition(modules,
+                                                         Constants.Security.Authorization.VISIBLE_BATCH_SIZE)) {
+      Map<DatasetModuleId, DatasetModuleMeta> datasetModulesMapping = new LinkedHashMap<>(split.size());
+      for (DatasetModuleMeta datasetModuleMeta : split) {
+        datasetModulesMapping.put(namespaceId.datasetModule(datasetModuleMeta.getName()), datasetModuleMeta);
+      }
+      datasetModulesMapping.keySet().retainAll(authorizationEnforcer.isVisible(datasetModulesMapping.keySet(),
+                                                                               principal));
+      visibleModules.addAll(Lists.newArrayList(datasetModulesMapping.values()));
+    }
+    return  visibleModules;
+  }
+
+  @Override
+  public DatasetModuleMeta getModule(DatasetModuleId datasetModuleId) throws Exception {
+    ensureOnePrivilege(datasetModuleId, EnumSet.of(Action.ADMIN, Action.READ));
+    return delegate.getModule(datasetModuleId);
+  }
+
+  @Override
+  public BodyConsumer addModule(DatasetModuleId datasetModuleId, String className,
+                                boolean forceUpdate) throws Exception {
+    final Principal principal = authenticationContext.getPrincipal();
+    // enforce that the principal has ADMIN access on the dataset module
+    authorizationEnforcer.enforce(datasetModuleId, principal, Action.ADMIN);
+    return delegate.addModule(datasetModuleId, className, forceUpdate);
+  }
+
+  @Override
+  public void delete(DatasetModuleId datasetModuleId) throws Exception {
+    Principal principal = authenticationContext.getPrincipal();
+    authorizationEnforcer.enforce(datasetModuleId, principal, Action.ADMIN);
+    delegate.delete(datasetModuleId);
+  }
+
+  @Override
+  public void deleteAll(NamespaceId namespaceId) throws Exception {
+    Principal principal = authenticationContext.getPrincipal();
+    for (DatasetModuleMeta meta : delegate.listModules(namespaceId)) {
+      DatasetModuleId datasetModuleId = namespaceId.datasetModule(meta.getName());
+      authorizationEnforcer.enforce(datasetModuleId, principal, Action.ADMIN);
+    }
+    delegate.deleteAll(namespaceId);
+  }
+
+  @Override
+  public List<DatasetTypeMeta> listTypes(final NamespaceId namespaceId) throws Exception {
+    List<DatasetTypeMeta> types = delegate.listTypes(namespaceId);
+
+    Principal principal = authenticationContext.getPrincipal();
+    List<DatasetTypeMeta> visibleTypes = new ArrayList<>(types.size());
+    for (List<DatasetTypeMeta> split : Lists.partition(types, Constants.Security.Authorization.VISIBLE_BATCH_SIZE)) {
```

Comment:
I can do it for `ensureOnePrivilege`, but for list, we need to construct the specific entity id from the info we get, how do we refactor this into one single helper method?